src/content/reference/device-os/firmware.md

```diff
@@ -12979,6 +13005,1754 @@ int8_t measurePower() const;
+bool has(const char* name) const;
+bool has(const String& name) const;
+```
+
+### isEmpty() [LedgerData class]
+
+Returns true if the ledger data is empty (has no entries).
+
+```cpp
+// PROTOTYPE
+bool isEmpty() const;
+```
+
+See [Variant](#variant) and [Map](#map) for additional information.
+
+
+## Variant
+
+{{api name1="Variant"}}
+
+{{since when="5.8.0"}}
+
+The `Variant` class holds typed data. It is used by [Ledger](#ledger). See also [VariantArray](#variantarray) and [VariantMap](#variantmap) to hold data that will be converted to JSON or CBOR.
+
+
+### set() [Variant class]
+
+One common use-case of Variant is setting values in a ledger. In this code, a fake sensor value is stored in `int sensorValue`. In a real application, you'd read an actual sensor there.
+
+Then a `Variant` object is created on the stack that will be filled with data.
+
+The `set()` method is described below in [VariantMap](#set-variantmap-). It takes a key name (such as "sensor" or "time") and a value, which is also a `Variant`. Because there are Variant constructors for many variable types (see [variant constructor](#constructor-variant-class-), below), you can pass a variable directly to set and the compiler will handle it appropriately. In the time example, the returned object from `Time.format` is a `String`.
+
+```cpp
+int sensorValue = readSensor();
+
+Variant data;
+data.set("sensor", sensorValue);
+data.set("time", Time.format(TIME_FORMAT_ISO8601_FULL));
+sensors.set(data);
+```
+
+See [Ledger sensor](/getting-started/logic-ledger/ledger-sensor/) for the full example.
+
+### Variant::Type
+
+Variants have an explicit type, unlike JSON. The following
+
+| Constant | C++ Type |
+| :--- | :--- |
+| Variant::Type::NULL | Null type (`std::monostate`) |
+| Variant::Type::BOOL | `bool` |
+| Variant::Type::INT | `int` |
+| Variant::Type::UINT | `unsigned` |
+| Variant::Type::INT64 | `int64_t` |
+| Variant::Type::UINT64 | `uint64_t` |
+| Variant::Type::DOUBLE | `double` |
+| Variant::Type::STRING | `String` |
+| Variant::Type::ARRAY | `VariantArray` |
+| Variant::Type::MAP | `VariantMap` |
+
+
+### constructor [Variant class]
+
+{{api name1="Variant::Variant()"}}
+
+You can construct a `Variant` object with a parameter of an explict type to create a variant of that type.
+
+```cpp
+// PROTOTYPES
+Variant();
+Variant(const std::monostate& val);
+Variant(bool val);
+Variant(int val);
+Variant(unsigned val);
+Variant(long val);
+Variant(unsigned long val);
+Variant(long val);
+Variant(unsigned long val);
+Variant(long long val);
+Variant(unsigned long long val);
+Variant(double val);
+Variant(const char* val);
+Variant(String val);
+Variant(VariantArray val);
+Variant(VariantMap val);'
+```
+
+### value() [Variant class]
+
+{{api name1="Variant::value()"}}
+
+The `value` method is a template that returns a reference to the stored value of the variant. The allowable types are listed in [Variant::Type](#variant-type). This can be done to either get or set a value.
```

Comment:
It should be documented that the template type of this method must match the actual underlying type of the variant value, otherwise the behavior is undefined. A safer alternative is to use the `toXXX()` and `asXXX()` methods.